data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/BulkRetryStrategy.java

```diff
@@ -152,16 +160,46 @@ public static boolean canRetry(final Exception e) {
                         !NON_RETRY_STATUS.contains(((OpenSearchException) e).status().getStatus())));
     }
 
+    private void handleRetriesAndFailures(final AccumulatingBulkRequest bulkRequestForRetry,
+                                          final int retryCount,
+                                          final BackOffUtils backOffUtils,
+                                          final BulkResponse bulkResponse,
+                                          Exception e) throws InterruptedException {
+        boolean doRetry = (Objects.isNull(e)) ? canRetry(bulkResponse) : canRetry(e);
+        if (!Objects.isNull(bulkResponse) && doRetry && retryCount == 1) { // first attempt
+            for (final BulkResponseItem bulkItemResponse : bulkResponse.items()) {
+                if (bulkItemResponse.error() == null) {
+                    sentDocumentsOnFirstAttemptCounter.increment();
+                }
+            }
+        }
+        if (doRetry && retryCount < maxRetries) {
+            handleRetry(bulkRequestForRetry, bulkResponse, backOffUtils);
+            bulkRequestNumberOfRetries.increment();
+        } else {
+            if (doRetry && retryCount >= maxRetries) {
+                e = new RuntimeException(String.format("Number of retries reached the limit of max retries(configured value %d)", maxRetries));
```

Comment:
We should avoid re-assignment of method parameters. This can be dangerous and leads to confusing code.

Reply:
In this case, the caller gets the new exception returned too, so that it gets passed back. Basically it is an in & out parameter to the function `handleRetriesAndFailures`